Tests/SwiftGRPCNIOTests/GRPCChannelHandlerTests.swift

```diff
@@ -0,0 +1,196 @@
+      let requestHead = HTTPRequestHead(version: .init(major: 2, minor: 0), method: .POST, uri: "unimplemented")
+      try channel.writeInbound(RawGRPCServerRequestPart.head(requestHead))
+    }
+
+    XCTAssertNoThrow(try extractStatus(responses[0])) { status in
+      XCTAssertEqual(status.code, .unimplemented)
+    }
+  }
+
+  func testImplementedMethodReturnsHeadersMessageAndStatus() throws {
+    let responses = try waitForGRPCChannelHandlerResponses(count: 3) { channel in
+      let requestHead = HTTPRequestHead(version: .init(major: 2, minor: 0), method: .POST, uri: "/echo.Echo/Get")
+      try channel.writeInbound(RawGRPCServerRequestPart.head(requestHead))
+
+      let request = Echo_EchoRequest.with { $0.text = "echo!" }
+      let requestData = try request.serializedData()
+      var buffer = channel.allocator.buffer(capacity: requestData.count)
+      buffer.write(bytes: requestData)
+      try channel.writeInbound(RawGRPCServerRequestPart.message(buffer))
+    }
+
+    XCTAssertNoThrow(try extractHeaders(responses[0]))
+    XCTAssertNoThrow(try extractMessage(responses[1]))
+    XCTAssertNoThrow(try extractStatus(responses[2])) { status in
+      XCTAssertEqual(status.code, .ok)
+    }
+  }
+
+  func testImplementedMethodReturnsStatusForBadlyFormedProto() throws {
+    let responses = try waitForGRPCChannelHandlerResponses(count: 2) { channel in
+      let requestHead = HTTPRequestHead(version: .init(major: 2, minor: 0), method: .POST, uri: "/echo.Echo/Get")
+      try channel.writeInbound(RawGRPCServerRequestPart.head(requestHead))
+
+      var buffer = channel.allocator.buffer(capacity: 3)
+      buffer.write(bytes: [1, 2, 3])
+      try channel.writeInbound(RawGRPCServerRequestPart.message(buffer))
+    }
+
+    XCTAssertNoThrow(try extractHeaders(responses[0]))
+    XCTAssertNoThrow(try extractStatus(responses[1])) { status in
+      let expectedStatus = GRPCStatus.requestProtoParseError
+      XCTAssertEqual(status.code, expectedStatus.code)
+      XCTAssertEqual(status.message, expectedStatus.message)
+    }
+  }
+}
+
+class HTTP1ToRawGRPCServerCodecTests: GRPCChannelHandlerResponseCapturingTestCase {
+  func testUnimplementedStatusReturnedWhenCompressionFlagIsSet() throws {
+    let responses = try waitForGRPCChannelHandlerResponses(count: 2) { channel in
+      let requestHead = HTTPRequestHead(version: .init(major: 2, minor: 0), method: .POST, uri: "/echo.Echo/Get")
+      try channel.writeInbound(HTTPServerRequestPart.head(requestHead))
+      try channel.writeInbound(HTTPServerRequestPart.body(gRPCMessage(channel: channel, compression: true)))
+    }
+
+    XCTAssertNoThrow(try extractHeaders(responses[0]))
+    XCTAssertNoThrow(try extractStatus(responses[1])) { status in
+      let expected = GRPCStatus.unsupportedCompression
+      XCTAssertEqual(status.code, expected.code)
+      XCTAssertEqual(status.message, expected.message)
+    }
+  }
+
+  func testMessageCanBeSentAcrossMultipleByteBuffers() throws {
```

Comment:
I'm curious — would this test have failed before this PR? I thought `HTTP1ToRawGRPCServerCodec` had already supported merging messages across frame boundaries, but I might be mistaken.

Reply:
It would, short of a couple of small issues; https://github.com/grpc/grpc-swift/blob/0195dfd5b3c7953dcff433a3eb86c27f0f5ff43a/Sources/SwiftGRPCNIO/HTTP1ToRawGRPCServerCodec.swift#L89 should have been `break requestProcessing` (otherwise you don't exit the while loop)

The other (which threw me for a while) was not reassigning to `self.buffer`, I _think_ what was happening is that when `buffer` was being written to here:

https://github.com/grpc/grpc-swift/blob/0195dfd5b3c7953dcff433a3eb86c27f0f5ff43a/Sources/SwiftGRPCNIO/HTTP1ToRawGRPCServerCodec.swift#L68

The underlying storage has references from `self.buffer` and `buffer`, so the storage was copied, we wrote into `buffer`s storage, so `self.buffer`s storage would still be empty.

Reply:
That makes sense, thank you for the fixes!